test/ad/distributions.jl

```diff
@@ -1,586 +1,621 @@
-        test_head("Testing: Univariate discrete distributions")
-        for d in uni_disc_dists
-            test_info(d.name)
-            for testf in get_all_functions(d, false)
-                test_ad(testf.f, testf.x)
-            end
+    @testset "Univariate distributions" begin
+        println("\nTesting: Univariate distributions\n")
 
-            dist = eval(:($(d.name)($(d.θ)...)))
-            x = d.x
-            y = link(dist, x)
-            @test invlink(dist, y) ≈ x
+        for d in univariate_distributions
+            test_ad(d)
         end
     end
-    separator()
-
-    # Note: broadcasting logpdf with univariate distributions having tracked parameters breaks
-    # Tracker. Ref: https://github.com/FluxML/Tracker.jl/issues/65
-    # filldist works around this so it is the recommended way for AD-friendly "broadcasting"
-    # of logpdf with univariate distributions
-
-    @testset "Univariate continuous distributions" begin
-        test_head("Testing: Univariate continuous distributions")
-        for d in uni_cont_dists
-            test_info(d.name)
-            for testf in get_all_functions(d, true)
-                test_ad(testf.f, testf.x)
-            end
 
-            dist = eval(:($(d.name)($(d.θ)...)))
-            x = d.x
-            y = link(dist, x)
-            @test invlink(dist, y) ≈ x
+    @testset "Multivariate distributions" begin
+        println("\nTesting: Multivariate distributions\n")
+
+        for d in multivariate_distributions
+            test_ad(d)
         end
-    end
-    separator()
-
-    @testset "Multivariate discrete distributions" begin
-        test_head("Testing: Multivariate discrete distributions")
-        for d in xmulti_disc_dists
-            test_info(d.name)
-            for testf in get_all_functions(d, false)
-                test_ad(testf.f, testf.x)
-            end
 
-            dist = eval(:($(d.name)($(d.θ)...)))
-            x = d.x
-            y = link(dist, x)
-            @test all(invlink(dist, y) .≈ x)
+        # Test `filldist` and `arraydist` distributions of univariate distributions
+        n = 2 # always use two distributions
+        for d in univariate_distributions
+            d.x isa Number || continue
+
+            # Broken distributions
+            d.f(d.θ...) isa Union{VonMises,TriangularDist} && continue
+
+            # Create `filldist` distribution
+            f_filldist = (θ...,) -> filldist(d.f(θ...), n)
+            d_filldist = f_filldist(d.θ...)
+
+            # Create `arraydist` distribution
+            f_arraydist = (θ...,) -> arraydist([d.f(θ...) for _ in 1:n])
+            d_arraydist = f_arraydist(d.θ...)
+
+            for sz in ((n,), (n, 2))
+                # Matrix case doesn't work for continuous distributions for some reason
+                # now but not too important (?!)
+                if length(sz) == 2 && Distributions.value_support(typeof(d)) === Continuous
+                    continue
+                end
```

Comment:
Do you know why? I.e. what's the error?

Reply:
No, I can check it. That's why I added "(?!)" to the comment :smile:

Reply:
BTW I think it's pretty bad to just state arrays of distributions for which tests are broken because we won't notice if some change (in Bijectors or some other package) fixes a broken test. IMO we should explicitly test whatever behaviour is broken. I just didn't want to put too many changes in this PR and wanted to check if it actually helps to reduce testing times first.

Reply:
It seems it's mainly an issue with Tracker (yields `MethodError: Cannot `convert` an object of type Array{Float64,1} to an object of type SubArray{Float64,1,Array{Float64,2},Tuple{Base.Slice{Base.OneTo{Int64}},Int64},true}` already for the first distribution), although Zygote also yields errors for some distributions such as `arraydist` with `GeneralizedExtremeValue` (but works fine for all distributions above, i.e., `Arcsine` to `Gamma`).

Reply:
Sounds like a good idea, but how would we do this in practice when checking PRs and stuff? Should have a one test-suite which tests broken tests and is allowed to fail? Doesn't seem like that will helps us discover if tests are fixed by upstream packages since there will be too much noise.

Reply:
I guess we would have to state explicitly for which AD backends tests don't pass, and then use `@test_broken` instead of `@test` for these inside of `test_ad`. Probably would need an additional argument to `DistSpec`, e.g., a field `broken` with default value `()` that could be set to `("Zygote",)`, `("Zygote", "ReverseDiff")` etc. Would be tedious to set up initially, but should work quite reliably I assume.

Reply:
Aaaah, pfft I haven't read the docs of `@test_broken` before; that makes complete sense. But should we maybe just get this PR merged now and then we can add this later? Unless you're _really_ keen on just getting it done immediately, in which case I won't stop you:)

Reply:
No, actually I'm not very keen on going through the lists and figuring out why these tests are broken right now :stuck_out_tongue: So I'm fine with adding this later.